storage/reader.go

```diff
@@ -392,6 +525,82 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := 0
+
+	if len(r.leftovers) > 0 {
+		cp := copy(p, r.leftovers)
+		r.seen += int64(cp)
+		r.leftovers = r.leftovers[cp:]
+
+		// No more room left in the user's buffer or the entire object has been
+		// read.
+		if len(p[cp:]) == 0 || r.size == r.seen {
+			return cp, nil
+		}
+
+		n = cp
+	}
+
+	for len(p[n:]) > 0 {
+		msg, err := r.stream.Recv()
+		if err != nil {
+			if err == io.EOF && r.seen == r.size {
+				// Let the next call to readWithGRPC return EOF.
+				return n, nil
+			}
+
+			// Do not retry PermissionDenied.
+			if st, ok := status.FromError(err); ok && st.Code() == codes.PermissionDenied {
+				return n, err
+			}
+
+			// TODO: How many times do we try to read before stopping?
```

Comment:
Presumably `ReadWithRetry` already has an answer to that question, shouldn't you use the same answer?

Reply:
@tritone can you point us to where this may be defined?

Reply:
Discussed this offline; retries will continue for retryable errors until the user cancels the context.

Reply:
Worked this out offline. Done!